### ISSUE TITLE
Design and approach shown for required fields mandatory

### ISSUE THREAD

**Issue by mcmark** [wet-boew/wet-boew#6760]
Are the design and approach shown for required fields on http://wet-boew.github.io/v4.0-ci/demos/formvalid/formvalid-en.html mandatory for all government web sites and applications moving to WET 4 or are they still just recommendations and departments can decide for themselves the approach they want to use for denoting mandatory fields.

My question also goes for the style guide as it relates to forms.   http://wet-boew.github.io/wet-boew-styleguide/v4/design/forms-en.html  Is the style guide mandatory for all government web sites and applications moving to WET 4 or is it just a recommendation and departments can decide for themselves the approach they want to use for styling forms and web pages in general.

Thanks,

Mark

**Comment by thomasgohard** [comment]
This really should be standardised across Canada.ca. @pcwsmith how about adding this to the spec?

**Comment by neoinsight** [comment]
@mcmark @pcwsmith For required fields, the design approach outlined for required fields in the Form demo was usability tested in January of this year and there were no issues, so I would support standardizing that approach, as seen below and in the demo. 
![form test jan2015](https://cloud.githubusercontent.com/assets/1855584/7822778/d02a36ce-03c5-11e5-8d1f-c51f05194426.png)

We were going to implement a demo form to capture the standardized approach for forms in general - the usability form is a good start at http://thomasgohard.github.io/web-form-validation-test/exform3-en.html but you need to know that some of the questions are there to compare with others, and therefore aren't recommended (e.g. auto-complete for country selection).

**Comment by hsrudnicki** [comment]
@mcmark @thomasgohard @neoinsight @LaurentGoderre : Can this issue be closed?

**Comment by pcwsmith** [comment]
@hsrudnicki please leave open.

As discussed yesterday with @thomasgohard @neoinsight @nrustad92 @lucas-hay @LaurentGoderre @dk-tbs, Thomas and team will remove (999-999-9999) from the WET form validation example. Proposed timeline: v4.0.16 or v4.0.17.

Rationale: Mobile users won’t/can’t enter number formats that are difficult to access via the mobile number pad. Forms need to accept input in a variety of forms (with/without hyphens, or parentheses, or spaces, etc.) given the range of devices/browsers in existence, not to mention the range of mental models that people have.

**Comment by pcwsmith** [comment]
Additionally, we discussed adding a postal code example to the WET form validation example page. Lisa from @neoinsight will provide her recommended specs.

**Comment by nschonni** [comment]
Telephone fields should probably using `type=tel` plus the `pattern` attribute so the mobile device will switch the input device to the numeric keyboard http://www.w3.org/wiki/HTML/Elements/input/tel
The support is pretty good depending on whether IE9 is still in scope http://caniuse.com/#search=tel

**Comment by neoinsight** [comment]
@nschonni Yes telephone fields absolutely should use `type=tel` That's the crux of this issue - asking people to use hyphens when the tel field pulls up a telephone dial pad without hyphens is unacceptable and in fact, unusable - mobile (and some desktop) participants did not enter the hyphens even when the formatting hint and placeholder hint to use hyphens were displayed. 

A `pattern` attribute is not recommended for tel fields for the reasons above  - the wiki you point to mentions that they should be free-form - it says 
>the Telephone type does not enforce a particular syntax. This is intentional; in practice, telephone number fields tend to be free-form fields, because there are a wide variety of valid phone numbers

**Comment by pjackson28** [comment]
PR #6952 adds a postal code example (as well as examples and/or documentation for the other validation methods with French translations). Also replaces the "(999-999-9999)" with "(including area code)" (since leaving out the area code will trigger an error message).

**Comment by pjackson28** [comment]
Note also that the telephone example already uses type="tel" (has for quite a while).

**Comment by neoinsight** [comment]
@pjackson28 - Terrific! Postal code field has 7 characters so they can put a space in if they want? and validates with either a space or without?

**Comment by pjackson28** [comment]
Umm... the postal code method from jQuery Validation (created by a GC department) is a little finicky. It wants upper-case letters and a space. Someone should probably fix this upstream. All my PR does is just adds a working example of what is already there.

@nschonni Any idea of how quickly could this be fixed upstream?

**Comment by nschonni** [comment]
It was create to follow the Canada Post guidelines strictly https://www.canadapost.ca/tools/pg/manual/PGaddress-e.asp

**Comment by pjackson28** [comment]
@nschonni Yes, but even Canada Post is not that strict when validating postal codes online. Those guidelines are for what gets printed. We need to be much more flexible at the data input stage and then can fix up the formatting as necessary after the postal code is already collected.

**Comment by pjackson28** [comment]
Canada Post seems to accept any combination of lowercase, uppercase, with space and without space.

**Comment by neoinsight** [comment]
@pjackson28 Great, Canada Post is doing it right then.  We just tested 20 people entering postal codes on the Service Canada site here: http://www.servicecanada.gc.ca/eng/sin/apply/how.shtml, and they did not use capital letters, and some used a space and some did not. Luckily, it worked for everyone, just like it should.

**Comment by pjackson28** [comment]
So it looks like the postal code validation will need to be fixed as we need to be able to support lowercase and no space (since it would otherwise be a 100% fail rate with those 20 users).

@masterbee @nschonni @thomasgohard Who can create the PR to fix the postal code validation upstream? How long will it be until the next jQuery validation version is released?

**Comment by hsrudnicki** [comment]
Due to no further discussion/status updates, this issue made "Going Stale" and closed as resolved based on the last PR merged.
If you strongly think of otherwise then please update this issue's status again, thank you

**Comment by QuesnelJ** [comment]
class="required" should not be tied to the strong tag but rather to a span tag. When a form has a mix of non-required and required fields, the labels end up not all having the same weight. In addition, there is no need to add emphasis to field labels. Emphasis should only be added when special attention is required.

**Comment by hsrudnicki** [comment]
@thomasgohard and others: Re-opened due to the last comment from @QuesnelJ  to resolve/complete fix or otherwise?

**Comment by kodecount** [comment]
There was furthered discussion here: https://github.com/wet-boew/wet-boew/issues/7896

This issue can be closed (instead of having multiple of the same issue).
